src/frequenz/dispatch/_dispatcher.py

```diff
@@ -358,9 +363,27 @@ async def __aenter__(self) -> Self:
             This background service.
         """
         await super().__aenter__()
+        await self._client.__aenter__()
         await self.wait_for_initialization()
         return self
 
+    async def stop(self, msg: str | None = None) -> None:
+        """Stop the local dispatch service and initiate client disconnection.
+
+        This method is called when the dispatcher is stopped.
+
+        Args:
+            msg: The message to log.
+        """
+        _logger.debug("Stopping dispatcher")
+        await self._bg_service.stop(msg)
+
+        for type in self._actor_dispatchers.keys():
+            await self.stop_managing(type)
+
+        await self._client.disconnect()
```

Comment:
Once you "fixed" `cancel()` and `wait()` there should be no need to override `stop()` too. In this case I guess you should move the actor stopping to `cancel()` too. And once you do that, maybe it makes sense to actually spawn a new task to do the cleanup. not sure, I guess you can also have a list of futures to wait for in the `wait()`, but you will not actually initiate the cleanup upon calling `cancel()`, only when calling `wait()`.

Reply:
Hmm so basically just remove stop() now? The actor stopping is already in cancel (using .cancel)..

Reply:
Then I guess so, I hope it works. Stopping/cleanup is pretty difficult/complicated at the moment, I hope we can improve things soon.